src/math.rs

```diff
@@ -68,6 +71,126 @@ impl Cube {
+        }
+    }
+    true
+}
+
+pub struct Obb<S> {
+    rotation_inv: Quaternion<S>,
+    translation_inv: Vector3<S>,
+    half_extent: Vector3<S>,
+    corners: [Point3<S>; 8],
+    separating_axes: Vec<Vector3<S>>,
+}
+
+impl<S: BaseFloat> Obb<S> {
+    pub fn new(rotation: Quaternion<S>, translation: Vector3<S>, half_extent: Vector3<S>) -> Self {
+        Obb {
+            rotation_inv: rotation.conjugate(),
+            translation_inv: -rotation.conjugate().rotate_vector(translation),
+            half_extent,
+            corners: Obb::precompute_corners(&rotation, &translation, &half_extent),
+            separating_axes: Obb::precompute_separating_axes(&rotation),
+        }
+    }
+
+    pub fn intersects(&self, aabb: &Aabb3<S>) -> bool {
+        intersects(&self.corners, &self.separating_axes, aabb)
+    }
+
+    pub fn contains(&self, p: &Point3<S>) -> bool {
+        let Point3 { x, y, z } = self.rotation_inv.rotate_point(*p) + self.translation_inv;
+        x.abs() <= self.half_extent.x
+            && y.abs() <= self.half_extent.y
+            && z.abs() <= self.half_extent.z
+    }
+
+    fn precompute_corners(
+        rotation: &Quaternion<S>,
+        translation: &Vector3<S>,
+        half_extent: &Vector3<S>,
+    ) -> [Point3<S>; 8] {
+        let transform =
+            |x: S, y: S, z: S| rotation.rotate_point(Point3::new(x, y, z)) + translation;
+        [
+            transform(-half_extent.x, -half_extent.y, -half_extent.z),
+            transform(half_extent.x, -half_extent.y, -half_extent.z),
+            transform(-half_extent.x, half_extent.y, -half_extent.z),
+            transform(half_extent.x, half_extent.y, -half_extent.z),
+            transform(-half_extent.x, -half_extent.y, half_extent.z),
+            transform(half_extent.x, -half_extent.y, half_extent.z),
+            transform(-half_extent.x, half_extent.y, half_extent.z),
+            transform(half_extent.x, half_extent.y, half_extent.z),
+        ]
+    }
+
+    fn precompute_separating_axes(rotation: &Quaternion<S>) -> Vec<Vector3<S>> {
+        let unit_x = Vector3::unit_x();
+        let unit_y = Vector3::unit_y();
+        let unit_z = Vector3::unit_z();
+        let rot_x = rotation.rotate_vector(unit_x);
+        let rot_y = rotation.rotate_vector(unit_y);
+        let rot_z = rotation.rotate_vector(unit_z);
+        let mut separating_axes = vec![unit_x, unit_y, unit_z];
+        for axis in &[
+            rot_x,
+            rot_y,
+            rot_z,
+            unit_x.cross(rot_x).normalize(),
+            unit_x.cross(rot_y).normalize(),
+            unit_x.cross(rot_z).normalize(),
+            unit_y.cross(rot_x).normalize(),
+            unit_y.cross(rot_y).normalize(),
+            unit_y.cross(rot_z).normalize(),
+            unit_z.cross(rot_x).normalize(),
+            unit_z.cross(rot_y).normalize(),
+            unit_z.cross(rot_z).normalize(),
+        ] {
+            let is_finite_and_non_parallel = is_finite(&axis)
```

Comment:
Is the non-parallel condition just an optimization?

Reply:
Yes, it's an optimization to get rid of as many axes to check before actually performing the intersection tests.